File: py/src/apptheory/kinesis_cloudwatch_logs.py
"""CloudWatch Logs subscription decoding for Lambda Kinesis records."""

from __future__ import annotations

import base64
import binascii
import gzip
import io
import json
import math
from typing import Any

CloudWatchLogsSubscription = dict[str, Any]
CloudWatchLogsSubscriptionLogEvent = dict[str, Any]
CloudWatchLogsSubscriptionSummary = dict[str, Any]

_CLOUDWATCH_LOGS_SUBSCRIPTION_MAX_DECODED_BYTES = 6 * 1024 * 1024
_CLOUDWATCH_LOGS_SUBSCRIPTION_DECODE_MESSAGE = "apptheory: decode cloudwatch logs subscription"


def decode_cloudwatch_logs_subscription(record: dict[str, Any]) -> CloudWatchLogsSubscription:
    """Decode a CloudWatch Logs subscription envelope from a Lambda Kinesis record.

    AWS delivers CloudWatch Logs subscription payloads to Kinesis as gzip-compressed JSON bytes encoded in the
    Lambda Kinesis record's ``kinesis.data`` field. The returned ``safe_summary`` intentionally excludes raw log event
    messages so callers can use it in logs, metrics, spans, and fixture summaries without copying customer log material.
    """

    record_obj = record if isinstance(record, dict) else {}
    record_id = _object_string(record_obj, "eventID")
    if not record_id:
        raise RuntimeError(f"{_CLOUDWATCH_LOGS_SUBSCRIPTION_DECODE_MESSAGE}: missing kinesis eventID")

    payload_bytes = _gunzip_cloudwatch_logs_subscription_data(record_obj)
    payload = _parse_cloudwatch_logs_subscription_payload(payload_bytes)
    decoded = {
        "record_id": record_id,
        "message_type": _object_string(payload, "messageType"),
        "owner": _object_string(payload, "owner"),
        "log_group": _object_string(payload, "logGroup"),
        "log_stream": _object_string(payload, "logStream"),
        "subscription_filters": _payload_string_array(payload, "subscriptionFilters"),
        "log_events": _payload_log_events(payload, "logEvents"),
    }
    _validate_cloudwatch_logs_subscription(decoded)
    decoded["safe_summary"] = _cloudwatch_logs_subscription_safe_summary(decoded)
    return decoded


def _gunzip_cloudwatch_logs_subscription_data(record: dict[str, Any]) -> bytes:
    kinesis = _object_from_value(record.get("kinesis"))
    data = kinesis.get("data")
    if isinstance(data, str):
        data_b64 = data.strip()
        if not data_b64:
            raise RuntimeError(f"{_CLOUDWATCH_LOGS_SUBSCRIPTION_DECODE_MESSAGE}: empty kinesis data")
        try:
            compressed = base64.b64decode(data_b64, validate=True)
        except (binascii.Error, ValueError) as exc:
            raise RuntimeError(f"{_CLOUDWATCH_LOGS_SUBSCRIPTION_DECODE_MESSAGE} gzip: invalid payload") from exc
    elif isinstance(data, bytes | bytearray | memoryview):
        compressed = bytes(data)
        if not compressed:
            raise RuntimeError(f"{_CLOUDWATCH_LOGS_SUBSCRIPTION_DECODE_MESSAGE}: empty kinesis data")
    else:
        raise RuntimeError(f"{_CLOUDWATCH_LOGS_SUBSCRIPTION_DECODE_MESSAGE}: empty kinesis data")

    try:
        with gzip.GzipFile(fileobj=io.BytesIO(compressed)) as reader:
            payload = reader.read(_CLOUDWATCH_LOGS_SUBSCRIPTION_MAX_DECODED_BYTES + 1)
    except (EOFError, OSError, ValueError) as exc:
        raise RuntimeError(f"{_CLOUDWATCH_LOGS_SUBSCRIPTION_DECODE_MESSAGE} gzip: invalid payload") from exc

    if len(payload) > _CLOUDWATCH_LOGS_SUBSCRIPTION_MAX_DECODED_BYTES:
        raise RuntimeError(f"{_CLOUDWATCH_LOGS_SUBSCRIPTION_DECODE_MESSAGE}: payload too large")
    return payload


def _parse_cloudwatch_logs_subscription_payload(payload_bytes: bytes) -> dict[str, Any]:
    try:
        parsed = json.loads(payload_bytes.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise RuntimeError(f"{_CLOUDWATCH_LOGS_SUBSCRIPTION_DECODE_MESSAGE} json: invalid payload") from exc
    if not isinstance(parsed, dict):
        raise RuntimeError(f"{_CLOUDWATCH_LOGS_SUBSCRIPTION_DECODE_MESSAGE} json: expected object")
    return parsed


def _validate_cloudwatch_logs_subscription(decoded: dict[str, Any]) -> None:
    missing: list[str] = []
    if not str(decoded.get("message_type") or "").strip():
        missing.append("messageType")
    if not str(decoded.get("owner") or "").strip():
        missing.append("owner")
    if not str(decoded.get("log_group") or "").strip():
        missing.append("logGroup")
    if not str(decoded.get("log_stream") or "").strip():
        missing.append("logStream")
    if not decoded.get("subscription_filters"):
        missing.append("subscriptionFilters")
    if missing:
        raise RuntimeError(f"{_CLOUDWATCH_LOGS_SUBSCRIPTION_DECODE_MESSAGE}: missing {', '.join(missing)}")

    for index, value in enumerate(decoded.get("subscription_filters") or []):
        if not str(value or "").strip():
            raise RuntimeError(f"{_CLOUDWATCH_LOGS_SUBSCRIPTION_DECODE_MESSAGE}: empty subscriptionFilters[{index}]")
    for index, event in enumerate(decoded.get("log_events") or []):
        if not str((event or {}).get("id") or "").strip():
            raise RuntimeError(f"{_CLOUDWATCH_LOGS_SUBSCRIPTION_DECODE_MESSAGE}: empty logEvents[{index}].id")


def _cloudwatch_logs_subscription_safe_summary(decoded: dict[str, Any]) -> CloudWatchLogsSubscriptionSummary:
    subscription_filter_count = len(decoded.get("subscription_filters") or [])
    log_event_count = len(decoded.get("log_events") or [])
    safe_log = (
        f"record_id={decoded.get('record_id')} owner={decoded.get('owner')} "
        f"log_group={decoded.get('log_group')} log_stream={decoded.get('log_stream')} "
        f"message_type={decoded.get('message_type')} log_events={log_event_count} "
        f"subscription_filters={subscription_filter_count}"
    )
    return {
        "record_id": str(decoded.get("record_id") or ""),
        "message_type": str(decoded.get("message_type") or ""),
        "owner": str(decoded.get("owner") or ""),
        "log_group": str(decoded.get("log_group") or ""),
        "log_stream": str(decoded.get("log_stream") or ""),
        "subscription_filter_count": subscription_filter_count,
        "log_event_count": log_event_count,
        "safe_log": safe_log,
    }


def _payload_string_array(payload: dict[str, Any], key: str) -> list[str]:
    value = payload.get(key)
    if not isinstance(value, list) or not value:
        return []
    return [item.strip() if isinstance(item, str) else "" for item in value]


def _payload_log_events(payload: dict[str, Any], key: str) -> list[CloudWatchLogsSubscriptionLogEvent]:
    value = payload.get(key)
    if not isinstance(value, list) or not value:
        return []
    out: list[CloudWatchLogsSubscriptionLogEvent] = []
    for item in value:
        event = _object_from_value(item)
        out.append(
            {
                "id": _object_string(event, "id"),
                "timestamp": _object_int(event, "timestamp"),
                "message": _object_raw_string(event, "message"),
            }
        )
    return out


def _object_from_value(value: Any) -> dict[str, Any]:
    return value if isinstance(value, dict) else {}


def _object_string(obj: dict[str, Any], key: str) -> str:
    value = obj.get(key) if isinstance(obj, dict) else ""
    return value.strip() if isinstance(value, str) else ""


def _object_raw_string(obj: dict[str, Any], key: str) -> str:
    value = obj.get(key) if isinstance(obj, dict) else ""
    return value if isinstance(value, str) else ""


def _object_int(obj: dict[str, Any], key: str) -> int:
    if not isinstance(obj, dict):
        return 0
    value = obj.get(key)
    if isinstance(value, bool):
        return 0
    if isinstance(value, int):
        return value
    if isinstance(value, float) and math.isfinite(value):
        return int(value)
    return 0

File: runtime/kinesis_cloudwatch_logs.go
package apptheory

import (
	"bytes"
	"compress/gzip"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"strings"

	"github.com/aws/aws-lambda-go/events"
)

const (
	cloudWatchLogsSubscriptionMaxDecodedBytes   = 6 * 1024 * 1024
	cloudWatchLogsSubscriptionDecodeMessage     = "apptheory: decode cloudwatch logs subscription"
	cloudWatchLogsSubscriptionDecodeGzipMessage = cloudWatchLogsSubscriptionDecodeMessage + " gzip: %w"
)

// CloudWatchLogsSubscription is a decoded CloudWatch Logs subscription envelope carried by a Kinesis record.
type CloudWatchLogsSubscription struct {
	RecordID            string                               `json:"record_id"`
	MessageType         string                               `json:"message_type"`
	Owner               string                               `json:"owner"`
	LogGroup            string                               `json:"log_group"`
	LogStream           string                               `json:"log_stream"`
	SubscriptionFilters []string                             `json:"subscription_filters"`
	LogEvents           []CloudWatchLogsSubscriptionLogEvent `json:"log_events"`
	SafeSummary         CloudWatchLogsSubscriptionSummary    `json:"safe_summary"`
}

// CloudWatchLogsSubscriptionLogEvent is one decoded CloudWatch Logs event from a subscription envelope.
type CloudWatchLogsSubscriptionLogEvent struct {
	ID        string `json:"id"`
	Timestamp int64  `json:"timestamp"`
	Message   string `json:"message"`
}

// CloudWatchLogsSubscriptionSummary is the safe, non-message summary for a decoded subscription envelope.
type CloudWatchLogsSubscriptionSummary struct {
	RecordID                string `json:"record_id"`
	MessageType             string `json:"message_type"`
	Owner                   string `json:"owner"`
	LogGroup                string `json:"log_group"`
	LogStream               string `json:"log_stream"`
	SubscriptionFilterCount int    `json:"subscription_filter_count"`
	LogEventCount           int    `json:"log_event_count"`
	SafeLog                 string `json:"safe_log"`
}

type cloudWatchLogsSubscriptionPayload struct {
	MessageType         string                               `json:"messageType"`
	Owner               string                               `json:"owner"`
	LogGroup            string                               `json:"logGroup"`
	LogStream           string                               `json:"logStream"`
	SubscriptionFilters []string                             `json:"subscriptionFilters"`
	LogEvents           []CloudWatchLogsSubscriptionLogEvent `json:"logEvents"`
}

// DecodeCloudWatchLogsSubscription decodes a Kinesis record containing a CloudWatch Logs subscription envelope.
//
// AWS delivers CloudWatch Logs subscription payloads to Kinesis as gzip-compressed JSON bytes. The returned
// SafeSummary intentionally excludes raw log event messages so callers can use it in logs, metrics, spans,
// and fixture summaries without copying customer log material.
func DecodeCloudWatchLogsSubscription(record events.KinesisEventRecord) (CloudWatchLogsSubscription, error) {
	decoded := CloudWatchLogsSubscription{RecordID: strings.TrimSpace(record.EventID)}
	if decoded.RecordID == "" {
		return decoded, errors.New(cloudWatchLogsSubscriptionDecodeMessage + ": missing kinesis eventID")
	}

	payloadBytes, err := gunzipCloudWatchLogsSubscriptionData(record.Kinesis.Data)
	if err != nil {
		return decoded, err
	}

	var payload cloudWatchLogsSubscriptionPayload
	if err := json.Unmarshal(payloadBytes, &payload); err != nil {
		return decoded, fmt.Errorf(cloudWatchLogsSubscriptionDecodeMessage+" json: %w", err)
	}

	decoded.MessageType = strings.TrimSpace(payload.MessageType)
	decoded.Owner = strings.TrimSpace(payload.Owner)
	decoded.LogGroup = strings.TrimSpace(payload.LogGroup)
	decoded.LogStream = strings.TrimSpace(payload.LogStream)
	decoded.SubscriptionFilters = trimCloudWatchLogsSubscriptionFilters(payload.SubscriptionFilters)
	decoded.LogEvents = cloneCloudWatchLogsSubscriptionLogEvents(payload.LogEvents)
	if err := validateCloudWatchLogsSubscription(decoded); err != nil {
		return decoded, err
	}
	decoded.SafeSummary = cloudWatchLogsSubscriptionSafeSummary(decoded)
	return decoded, nil
}

func gunzipCloudWatchLogsSubscriptionData(data []byte) ([]byte, error) {
	if len(data) == 0 {
		return nil, errors.New(cloudWatchLogsSubscriptionDecodeMessage + ": empty kinesis data")
	}
	reader, err := gzip.NewReader(bytes.NewReader(data))
	if err != nil {
		return nil, fmt.Errorf(cloudWatchLogsSubscriptionDecodeGzipMessage, err)
	}

	limited := io.LimitReader(reader, cloudWatchLogsSubscriptionMaxDecodedBytes+1)
	payload, readErr := io.ReadAll(limited)
	closeErr := reader.Close()
	if readErr != nil {
		return nil, fmt.Errorf(cloudWatchLogsSubscriptionDecodeGzipMessage, readErr)
	}
	if closeErr != nil {
		return nil, fmt.Errorf(cloudWatchLogsSubscriptionDecodeGzipMessage, closeErr)
	}
	if len(payload) > cloudWatchLogsSubscriptionMaxDecodedBytes {
		return nil, errors.New(cloudWatchLogsSubscriptionDecodeMessage + ": payload too large")
	}
	return payload, nil
}

func validateCloudWatchLogsSubscription(decoded CloudWatchLogsSubscription) error {
	var missing []string
	if strings.TrimSpace(decoded.MessageType) == "" {
		missing = append(missing, "messageType")
	}
	if strings.TrimSpace(decoded.Owner) == "" {
		missing = append(missing, "owner")
	}
	if strings.TrimSpace(decoded.LogGroup) == "" {
		missing = append(missing, "logGroup")
	}
	if strings.TrimSpace(decoded.LogStream) == "" {
		missing = append(missing, "logStream")
	}
	if len(decoded.SubscriptionFilters) == 0 {
		missing = append(missing, "subscriptionFilters")
	}
	if len(missing) > 0 {
		return fmt.Errorf(cloudWatchLogsSubscriptionDecodeMessage+": missing %s", strings.Join(missing, ", "))
	}

	for i, filter := range decoded.SubscriptionFilters {
		if strings.TrimSpace(filter) == "" {
			return fmt.Errorf(cloudWatchLogsSubscriptionDecodeMessage+": empty subscriptionFilters[%d]", i)
		}
	}
	for i, event := range decoded.LogEvents {
		if strings.TrimSpace(event.ID) == "" {
			return fmt.Errorf(cloudWatchLogsSubscriptionDecodeMessage+": empty logEvents[%d].id", i)
		}
	}
	return nil
}

func trimCloudWatchLogsSubscriptionFilters(in []string) []string {
	if len(in) == 0 {
		return nil
	}
	out := make([]string, len(in))
	for i, value := range in {
		out[i] = strings.TrimSpace(value)
	}
	return out
}

func cloneCloudWatchLogsSubscriptionLogEvents(in []CloudWatchLogsSubscriptionLogEvent) []CloudWatchLogsSubscriptionLogEvent {
	if len(in) == 0 {
		return nil
	}
	out := make([]CloudWatchLogsSubscriptionLogEvent, len(in))
	for i, event := range in {
		event.ID = strings.TrimSpace(event.ID)
		out[i] = event
	}
	return out
}

func cloudWatchLogsSubscriptionSafeSummary(decoded CloudWatchLogsSubscription) CloudWatchLogsSubscriptionSummary {
	filterCount := len(decoded.SubscriptionFilters)
	logEventCount := len(decoded.LogEvents)
	safeLog := fmt.Sprintf(
		"record_id=%s owner=%s log_group=%s log_stream=%s message_type=%s log_events=%d subscription_filters=%d",
		decoded.RecordID,
		decoded.Owner,
		decoded.LogGroup,
		decoded.LogStream,
		decoded.MessageType,
		logEventCount,
		filterCount,
	)
	return CloudWatchLogsSubscriptionSummary{
		RecordID:                decoded.RecordID,
		MessageType:             decoded.MessageType,
		Owner:                   decoded.Owner,
		LogGroup:                decoded.LogGroup,
		LogStream:               decoded.LogStream,
		SubscriptionFilterCount: filterCount,
		LogEventCount:           logEventCount,
		SafeLog:                 safeLog,
	}
}
